### PR TITLE
fix flaws when building

### PR DESCRIPTION
**Pull request by peterbe**

### CODE REVIEW

File: build/index.js [hunk 1]
```diff
@@ -322,7 +320,20 @@ async function buildDocuments() {
...
 }
 
+// function poorMansCLI() {
+//   require("dotenv").config();
+//   // XXX We need some form of CLI that can override and amend the `options`
+//   // object imported from ./build-options.js
+//   const argv = process.argv.slice(2);
+//   console.log("BUILD OPTIONS ARE NOW:");
+//   console.log(options);
+// }
```
Comment (Gregoor) [inline build/index.js:330]: nit: dead code police 👮‍♀️

**Review by Gregoor — APPROVED** [review]
I'm still a little confused with flaws wrt who wants to use/configure them how. It's another area where I'd err on the side of less configurability and defaults (e.g. for now: warn - in local mode, error - CI for new flaws, ignore - when prod building). But I might not've made my homework here, so if you think this is the way to go, let's do it.

**Comment by peterbe** [comment]
> I'm still a little confused with flaws wrt who wants to use/configure them how.

Who isn't?!

A lot of my "design decisions" are based on an early exploration about how we might want to use this. If in doubt, I start with making it flexible. Is it important to be able to have *different levels* of different types of flaws? I'm not sure. Is important to able to do things differently depending on `yarn build` vs. `yarn start`? Yes, but nothing that can't be accomplished with env vars. 

Here's what's somewhat firm:

* When doing prod builds, we definitely don't want to do any flaw checking. They're potentially (unnecessarily) expensive and they'd put `.flaws` in the `index.json` where they are useless. 
* In PR builds, in CI, we probably want to default to `warn` so that we could potentially either A) break the build if the flaws go up, or B) post a comment on the PR that says "FYI, the documents you touched have the following flaws..."
* In local development, for a writer, you probably don't distinguish between different flaws. They're all something you ought to worry about. 
* Someday, when the number of flaws reaches 0, we'll want the PR builds to build with flaw level `error` so that any introduction of a new flaw is bad enough to send the PR author back to the drawing board. (But note, this might be tricky because of redirects and broken links)
* During development (for the dev team) it's proven useful to be able to set `macros:error, *:warn` so that no KS macro goes unnoticed. 

Anything else you can share @escattone ?

**Comment by Gregoor** [comment]
Thanks for the explainer! The wildcard DSL thing makes me think you're expecting a lot more types of flaws. Given what's currently there I'd also consider a good ol' JSON config which would make `checkFlawLevels` (bit of a misnomer imo, should be something like `parseFlawLevels`) something like:
```javascript
const CUSTOM_CONFIG_PATH = "flaw-levels.json";

const DEFAULT_CONFIG = {
  macros: "warn",
  broken_links: "warn",
  bad_bcd_queries: "warn",
};

function parseFlawLevels() {
  const userConfig = fs.existsSync(CUSTOM_CONFIG_PATH)
    ? JSON.parse(fs.readFileSync(CUSTOM_CONFIG_PATH))
    : {};

  // if we're feeling fancy
  const unknownKeys = Object.keys(userConfig).filter(
    (key) => !Object.keys(DEFAULT_CONFIG).includes(key)
  );
  if (unknownKeys.length > 0) {
    throw new Error(
      `Your ${CUSTOM_CONFIG_PATH} config contains unknown keys: ${JSON.stringify(keys)}`
    );
  }

  return {
    ...DEFAULT_CONFIG,
    ...userConfig,
  };
}
```

Just a thought, and like I said. Might be less future-proof.

**Comment by peterbe** [comment]
I definitely think we should keep an eye on a solution like that! It might make a munch buncha sense in some future where there are LOTs of flaw checks and maybe even they can be plugins you write on your personal computer. 
But a path to a config file comes with its own set of challenges. Given the situation and context at the moment, I think I'm happy enough with the current env var + `checkFlawLevels` function. Partly because it's been coded now already and I'd weigh progress over perfect at the current time. 

However, one thing I like about your above code snippet is that the `DEFAULT_CONFIG`'s `.keys()` is the list of possible keys so you wouldn't need a separate `VALID_FLAW_CHECKS`. But I think we can revisit it later.

**Comment by peterbe** [comment]
I extended things so it can now do other options which are very relevant for mass-builds: https://github.com/Gregoor/yari/pull/5

**Pull request closed without merging**